File: frontend/src/types/relation.ts
export interface RoleRead {
  id: string;
  relation_revision_id: string;
  entity_id: string;
  role_type: string;
  weight?: number | null;
  coverage?: number | null;
  entity_slug?: string | null;
  disagreement?: number | null;
}

export interface RelationRevisionRead {
  id: string;
  relation_id: string;
  kind?: string | null;
  direction?: string | null;
  confidence?: number | null;
  scope?: Record<string, unknown> | null;
  evidence_context?: Record<string, unknown> | null;
  notes?: string | Record<string, string> | null;
  created_with_llm?: string | null;
  created_by_user_id?: string | null;
  created_at: string;
  is_current: boolean;
  status: string;
  llm_review_status?: string | null;
  roles: RoleRead[];
}

export interface RelationRead {
  id: string;
  created_at: string;
  updated_at: string;
  source_id: string;
  source_title?: string | null;
  source_year?: number | null;
  kind?: string | null;
  direction?: string | null;
  confidence?: number | null;
  scope?: Record<string, unknown> | null;
  evidence_context?: Record<string, unknown> | null;
  notes?: string | Record<string, string> | null;
  created_with_llm?: string | null;
  status: "draft" | "confirmed";
  llm_review_status?: string | null;
  roles: RoleRead[];
}

File: backend/app/schemas/relation.py
from uuid import UUID
from typing import List, Optional
from datetime import datetime
from app.schemas.base import Schema


class RoleRevisionWrite(Schema):
    """Schema for writing a role within a relation revision."""
    entity_id: UUID
    role_type: str
    weight: Optional[float] = None  # For computed relations
    coverage: Optional[float] = None  # For computed relations


class RoleRevisionRead(RoleRevisionWrite):
    """Schema for reading a role revision."""
    id: UUID
    relation_revision_id: UUID
    entity_slug: Optional[str] = None  # Resolved entity slug for display


class RelationWrite(Schema):
    """
    Schema for creating a new relation.

    Creates both the base Relation and its first RelationRevision.
    """
    source_id: UUID  # Immutable - relation always from one source
    kind: Optional[str] = None
    direction: Optional[str] = None
    confidence: Optional[float] = None
    scope: Optional[dict] = None  # Contextual qualifiers
    notes: Optional[dict] = None  # i18n: {"en": "...", "fr": "..."}
    roles: List[RoleRevisionWrite]
    created_with_llm: Optional[str] = None


class RelationRevisionRead(Schema):
    """Schema for reading a relation revision."""
    id: UUID
    relation_id: UUID
    kind: Optional[str] = None
    direction: Optional[str] = None
    confidence: Optional[float] = None
    scope: Optional[dict] = None
    notes: Optional[dict] = None
    created_with_llm: Optional[str] = None
    created_by_user_id: Optional[UUID] = None
    created_at: datetime
    is_current: bool
    roles: List[RoleRevisionRead]


class RelationRead(Schema):
    """
    Schema for reading a relation with its current revision.

    Combines base relation + current revision for convenience.
    """
    id: UUID
    created_at: datetime
    source_id: UUID  # Immutable

    # Current revision data
    kind: Optional[str] = None
    direction: Optional[str] = None
    confidence: Optional[float] = None
    scope: Optional[dict] = None
    notes: Optional[dict] = None
    roles: List[RoleRevisionRead]


class RelationWithHistory(RelationRead):
    """Relation with full revision history."""
    revisions: List[RelationRevisionRead] = []